### ISSUE TITLE
Che workspace creation process and Git repo cloning take too much time

### ISSUE THREAD

**Issue by deanone** [eclipse-researchlabs/smartclide-ide-front-end#13]
Dear @goncalorolo @philipreimer, @nikosnikolaidis 

I am testing the new release of our platform, by creating new test services. I have noticed that the following two processes take too much time:

1. Che workspace creation process
2. Git repo cloning process

Is it normal for these processes to take too much time? @goncalorolo I have added you to this issue as responsible for the frontend development process, @philipreimer I have added you since you are managing the deployment of our platform to the cluster, and @nikosnikolaidis I have added you because you are responsible for the service-creation backend service.

Can you please check and tell me if I am supposed to wait for too long for these processes to finish? Especially the cloning process, seem to go on forever (more than 5 minutes!).

Best,

Athanasios

**Comment by goncalorolo** [comment]
With "Che workspace creation" you mean the actual creation process (since you press "Create" [service] until you are shown the "Service created" message)? Or do you mean the process of running the workspace for the first time?

**Comment by nikosnikolaidis** [comment]
Dear @deanone

I just tested the Service Creation workflow.
The Service Creation service and Git clone should be very fast.
The issue seems to be that the first time the user should provide his/her git credentials in Theia as well.
At the bottom right the following message appears, where the user for the first time needs to provide the username and password. I don't know if we can provide these information in the Dev file, so that the user shouldn't be asked for a second time. Maybe @goncalorolo knows?
![image](https://user-images.githubusercontent.com/12414234/191020552-fe6576d8-7d32-4637-a14e-32b4c64a0c39.png)

**Comment by deanone** [comment]
> With "Che workspace creation" you mean the actual creation process (since you press "Create" [service] until you are shown the "Service created" message)? Or do you mean the process of running the workspace for the first time?

@philipreimer I mean running the workspace for the first time. For example, I create the service, and I do not hit the code icon to launch the Che workspace for the first time. Then, I go to the list of service, I find my newly created service, and I hit the code icon to launch the Che workspace. This very first launch of the Che workspace takes some time. Is it normal, considering that it sets up all dependencies required (e.g., nodejs installation or python installation)?

**Comment by deanone** [comment]
> Dear @deanone
> 
> I just tested the Service Creation workflow. The Service Creation service and Git clone should be very fast. The issue seems to be that the first time the user should provide his/her git credentials in Theia as well. At the bottom right the following message appears, where the user for the first time needs to provide the username and password. I don't know if we can provide these information in the Dev file, so that the user shouldn't be asked for a second time. Maybe @goncalorolo knows? ![image](https://user-images.githubusercontent.com/12414234/191020552-fe6576d8-7d32-4637-a14e-32b4c64a0c39.png)

Dear @nikosnikolaidis , yes the first time cloning takes too much time. But normally, cloning should take place (almost) instantly. Since the user has already provided the Git credentials in his profile set up, it does not make sense to be prompted again for them.

**Comment by goncalorolo** [comment]
> > With "Che workspace creation" you mean the actual creation process (since you press "Create" [service] until you are shown the "Service created" message)? Or do you mean the process of running the workspace for the first time?
> 
> @philipreimer I mean running the workspace for the first time. For example, I create the service, and I do not hit the code icon to launch the Che workspace for the first time. Then, I go to the list of service, I find my newly created service, and I hit the code icon to launch the Che workspace. This very first launch of the Che workspace takes some time. Is it normal, considering that it sets up all dependencies required (e.g., nodejs installation or python installation)?

Now that I am sure I know what you are referring to, yes, it does take some time when you run a workspace for the first time. It is not very user-friendly but there is nothing we can do about that, at least from our side.

**Comment by goncalorolo** [comment]
> > Dear @deanone
> > I just tested the Service Creation workflow. The Service Creation service and Git clone should be very fast. The issue seems to be that the first time the user should provide his/her git credentials in Theia as well. At the bottom right the following message appears, where the user for the first time needs to provide the username and password. I don't know if we can provide these information in the Dev file, so that the user shouldn't be asked for a second time. Maybe @goncalorolo knows? ![image](https://user-images.githubusercontent.com/12414234/191020552-fe6576d8-7d32-4637-a14e-32b4c64a0c39.png)
> 
> Dear @nikosnikolaidis , yes the first time cloning takes too much time. But normally, cloning should take place (almost) instantly. Since the user has already provided the Git credentials in his profile set up, it does not make sense to be prompted again for them.

Regarding the cloning, I have just tested and, aside from the fact that you have to provide the credentials (again) when you reach Che-Theia, the cloning was pretty much instantaneous.

Nonetheless, I am checking the possibility to provide the credentials in the devfile in order to prevent this behaviour.

**Comment by deanone** [comment]
@goncalorolo ok I guess we cannot do much for the time required to set up and launch a Che workspace for the first time.

Regarding cloning, yes it would be much better from a user perspective is the user was not prompt again to provide his credentials. I live the issue opened until you manage to put the user Git credentials into the devfile.

Thanks.

**Comment by philipreimer** [comment]
> @philipreimer I mean running the workspace for the first time. For example, I create the service, and I do not hit the code icon to launch the Che workspace for the first time. Then, I go to the list of service, I find my newly created service, and I hit the code icon to launch the Che workspace. This very first launch of the Che workspace takes some time. Is it normal, considering that it sets up all dependencies required (e.g., nodejs installation or python installation)?

I assume this is the same as when you start one of the sample workspaces provided by Che? For example, when you go to https://che-smartclide-che.che.smartclide.eu/dashboard/#/create-workspace and then just click on any of the demo workspaces under "Select a sample".
This initial start takes quite some time. You can see in the "Logs" tab then that Che pulls and starts a few container images. And this apparently takes some time.

**Comment by deanone** [comment]
> > @philipreimer I mean running the workspace for the first time. For example, I create the service, and I do not hit the code icon to launch the Che workspace for the first time. Then, I go to the list of service, I find my newly created service, and I hit the code icon to launch the Che workspace. This very first launch of the Che workspace takes some time. Is it normal, considering that it sets up all dependencies required (e.g., nodejs installation or python installation)?
> 
> I assume this is the same as when you start one of the sample workspaces provided by Che? For example, when you go to https://che-smartclide-che.che.smartclide.eu/dashboard/#/create-workspace and then just click on any of the demo workspaces under "Select a sample". This initial start takes quite some time. You can see in the "Logs" tab then that Che pulls and starts a few container images. And this apparently takes some time.

Yes, I understand. So, there is nothing we can do about it (for now). So, let's leave this issue open until the Git credentials are included to the devfile.

**Comment by goncalorolo** [comment]
> Yes, I understand. So, there is nothing we can do about it (for now). So, let's leave this issue open until the Git credentials are included to the devfile.

Just to give an update on this, I found this guide: https://www.eclipse.org/che/docs/stable/administration-guide/configuring-oauth-2-for-gitlab/#setting-up-the-gitlab-authorized-application_che. It seems that this type of configuration involves creating the so-called Secret. The problem is that some actions on the deployment side are required. Hence, I am not seeing how we could automate this for each service that the user creates.

**Comment by deanone** [comment]
> > Yes, I understand. So, there is nothing we can do about it (for now). So, let's leave this issue open until the Git credentials are included to the devfile.
> 
> Just to give an update on this, I found this guide: https://www.eclipse.org/che/docs/stable/administration-guide/configuring-oauth-2-for-gitlab/#setting-up-the-gitlab-authorized-application_che. It seems that this type of configuration involves creating the so-called Secret. The problem is that some actions on the deployment side are required. Hence, I am not seeing how we could automate this for each service that the user creates.

@goncalorolo yes it seems a little tricky. @philipreimer do you think that this "Applying the GitLab-authorized application Secret" part of the process can be automated somehow?

**Comment by goncalorolo** [comment]
> > Yes, I understand. So, there is nothing we can do about it (for now). So, let's leave this issue open until the Git credentials are included to the devfile.
> 
> Just to give an update on this, I found this guide: https://www.eclipse.org/che/docs/stable/administration-guide/configuring-oauth-2-for-gitlab/#setting-up-the-gitlab-authorized-application_che. It seems that this type of configuration involves creating the so-called Secret. The problem is that some actions on the deployment side are required. Hence, I am not seeing how we could automate this for each service that the user creates.

Allow me to add that this may not solve the issue though. The main point is that it does not seem possible to send the credentials via devfile.

**Comment by deanone** [comment]
> > > Yes, I understand. So, there is nothing we can do about it (for now). So, let's leave this issue open until the Git credentials are included to the devfile.
> > 
> > 
> > Just to give an update on this, I found this guide: https://www.eclipse.org/che/docs/stable/administration-guide/configuring-oauth-2-for-gitlab/#setting-up-the-gitlab-authorized-application_che. It seems that this type of configuration involves creating the so-called Secret. The problem is that some actions on the deployment side are required. Hence, I am not seeing how we could automate this for each service that the user creates.
> 
> Allow me to add that this may not solve the issue though. The main point is that it does not seem possible to send the credentials via devfile.

Any other way? I mean apart from the devfile.

**Comment by goncalorolo** [comment]
> @goncalorolo any other way? I mean apart from the devfile.

Nothing for now...

**Comment by deanone** [comment]
Ok, @goncalorolo let's leave it open and investigate it a little bit further.
